homeassistant/components/template/template_entity.py

```diff
@@ -430,3 +445,220 @@ async def async_added_to_hass(self) -> None:
     async def async_update(self) -> None:
         """Call for forced update."""
         self._async_update()
+
+
+class TemplateRestoreEntity(RestoreEntity, TemplateEntity):
+    """Template Entity that restores data."""
+
+    def __init__(self, *args, **kwargs) -> None:
+        """Template Restore Entity init."""
+        super().__init__(*args, **kwargs)
+        self._restore: bool = False
+        self._save_state: bool = True
+        self._additional_data: list[str] = []
+
+    def add_template_attribute(
+        self, attribute: str, template: Template, *args, **kwargs
+    ):
+        """Store attribute to allow restore and setup template."""
+
+        super().add_template_attribute(attribute, template, *args, **kwargs)
+        if not template.is_static:
+            self.add_additional_data(attribute)
+
+    @callback
+    def _handle_results(self, *args, **kwargs) -> None:
+        # Do not save state if handle result fails.
+        self._save_state = False
+        super()._handle_results(*args, **kwargs)
+        self._save_state = True
+
+    @property
+    def restore(self) -> bool:
+        """Retrieve restore."""
+        return self._restore or False
+
+    @restore.setter
+    def restore(self, restore: bool) -> None:
+        """Set restore."""
+        self._restore = restore
+
+    @property
+    def save_state(self) -> bool:
+        """Retrieve save state."""
+        return self._save_state
+
+    @property
+    def additional_data(self) -> list[str]:
+        """Return additional data list."""
+        return self._additional_data or []
+
+    def add_additional_data(self, attribute: str) -> None:
+        """Add attribute to additional data list."""
+        if not hasattr(self, "_additional_data"):
```

Comment:
This is impossible since we set the attribute in the initializer?

Reply:
This is developed to support all the different template platforms but this initial PR is for sensor platform with the idea that once this PR is completed I will then submit other PRs to include the other platforms. 
But this PR lays the framework for the other platforms as well. Some of those inherit from RestoreEntity already and thus won't end up calling the __init__ for class TemplateRestoreEntity. That is why I check if the attribute exist as it might not.